### ISSUE TITLE
Channel creation/destruction may leave dangling thread.

### ISSUE THREAD

**Issue by meawoppl** [grpc/grpc#14338]
### What version of gRPC and what language are you using?
- Python==3.5.2
- grpcio==1.8.4
- grpcio-tools==1.8.4
 
### What operating system (Linux, Windows, …) and version?
 - Linux Ubuntu 16.04
 - Windows 7
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 - Python==3.5.2

### What did you do?

Here is an example of a minimal function call we wrote
```Python
def wait_for_channel_ready(channel: grpc.Channel, timeout_ms: numbers.Number):
    """
    Wait for `channel` to enter the ready state.  If it does not enter this state
    withing `timeout_seconds`, raise a TimeoutError.

    :param channel: a grcp.Channel instance.
    :param timeout_ms: Time in milliseconds to wait for Channel to be ready.
    """
    fut = grpc.channel_ready_future(channel)
    try:
        fut.result(timeout=timeout_ms / 1000)
    except grpc.FutureTimeoutError as e:
        fut.cancel()
        raise TimeoutError(
            "Timeout waiting on {} for {} ms".format(channel, timeout_ms)) from e
```
And here is the corresponding test:
```Python
def test_waiter_times_out():
     channel = grpc.insecure_channel("localhost:51234")  # No one answers here
     with nose.tools.assert_raises(TimeoutError):
        grpc_helpers.wait_for_channel_ready(channel, 5)

    # This tries to trigger the __del__ handlers that grpc has decided
    # actually work in Python:  https://github.com/grpc/grpc/issues/12531
    del channel
    gc.collect()
```

### What did you expect to see?

Test passes leaving no lingering non-demonic threads around.

### What did you see instead?
Following the above the context in which it was run will have accrued a unterminated non-demonic thread.  Code like the following (enforced by our CI) will fail:
```Python
import threading
for t in threading.enumerate():
   assert t.daemon, t
```

**Our CI requires that tests do not leave Threads of this nature after the test has completed**, and it will  flags tests that leave unterminated threads around as having failed failed.  It seems there is no (documented) way to terminate these lingering threads.   
 
### Anything else we should know about your project / environment?
I believe this is linked to #12531.  The behavior is intermittent, and does not fail in all cases.  Including sleep/timeout retry logic does not seem to help.

**Comment by meawoppl** [comment]
Possibly also #6666

**Comment by nathanielmanistaatgoogle** [comment]
I also believe that this is linked to [issue 12531](https://github.com/grpc/grpc/issues/12531); let's reassess after that issue is closed.

**Comment by nathanielmanistaatgoogle** [comment]
@meawoppl: if you are able to work from source, it may be time to reassess this issue now that [issue 12531](https://github.com/grpc/grpc/issues/12531) is closed. Otherwise the [issue 12531](https://github.com/grpc/grpc/issues/12531) fix will be included in the upcoming 1.12 release, so reassess this issue then?

**Comment by meawoppl** [comment]
We are currently pinned to 1.11, when there is a 1.12 release it would be easy for me to test.  If you ping back on this thread at that time, I would be happy to confirm/deny whether this got fixed.

**Comment by nathanielmanistaatgoogle** [comment]
[`grpcio` 1.12.0 was uploaded about an hour ago](https://pypi.org/project/grpcio/1.12.0); please start working with it. In particular a `grpc.Channel.close` method is introduced; we'd be very keen to hear whether or not deliberately closing your channels resolves this issue.

**Comment by nathanielmanistaatgoogle** [comment]
@meawoppl: poke poke? Does the problem continue to manifest? Should this issue remain open?

**Comment by meawoppl** [comment]
I gave this a shot today and traded flaky thread cleanup for segfaults.  The test suite appears to have rotted a bit, so more soon.

**Comment by meawoppl** [comment]
Updates in #12531 should I repost them here too?

**Comment by srini100** [comment]
No needed. These recent fixes on master and available in upcoming 1.13.0 should resolve your segfault on shutdown issue. Well, hopefully. Update us when you get a chance to try these out.

https://github.com/grpc/grpc/pull/15825
https://github.com/grpc/grpc/pull/15727

**Comment by srini100** [comment]
Since 1.13.0 has been out for a while, closing this issue. Please reopen if the issue still exits.

**Comment by meawoppl** [comment]
Taking a poke at 1.13.0 this week. Will open a new ticket if this problem
continues.

**Comment by nathanielmanistaatgoogle** [comment]
@meawoppl: if you see the problem with the latest release and it's still the same problem described here, reopen this issue rather than opening a new issue.
